csi/controller.go

```diff
@@ -455,9 +471,122 @@ func csiRequestsSharedVolume(req *csi.CreateVolumeRequest) bool {
 	return false
 }
 
-/*
-For next patches what still needs to be worked on in the Conroller server:
+// CreateSnapshot is a CSI implementation to create a snapshot from the volume
+func (o *OsdCsiServer) CreateSnapshot(
+	ctx context.Context,
+	req *csi.CreateSnapshotRequest,
+) (*csi.CreateSnapshotResponse, error) {
 
-	GetCapacity(context.Context, *GetCapacityRequest) (*GetCapacityResponse, error)
-	ControllerGetCapabilities(context.Context, *ControllerGetCapabilitiesRequest) (*ControllerGetCapabilitiesResponse, error)
-*/
+	if len(req.GetSourceVolumeId()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Volume id must be provided")
+	} else if len(req.GetName()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Name must be provided")
+	}
+
+	// Check if the snapshot with this name already exists
+	v, err := util.VolumeFromName(o.driver, req.GetName())
```

Comment:
A create call should fail if the resource already exists.

This is keeping inline with behavior of other systems like REST POSTs, kubectl creates, docker volume create...

Reply:
@harsh-px CSI Spec defines that all calls are idempotent. If the create call has a name that has already been taken, then the spec specifies that the function must return the ID of that snapshot